src/equations/ideal_glm_mhd_multiion.jl

```diff
@@ -0,0 +1,256 @@
+
+    for k in eachcomponent(equations)
+        rho, rho_v1, rho_v2, rho_v3, _ = get_component(k, u, equations)
+
+        total_electron_charge += rho * equations.charge_to_mass[k]
+        vk1_plus[k] = rho_v1 * equations.charge_to_mass[k]
+        vk2_plus[k] = rho_v2 * equations.charge_to_mass[k]
+        vk3_plus[k] = rho_v3 * equations.charge_to_mass[k]
+    end
+    vk1_plus ./= total_electron_charge
+    vk2_plus ./= total_electron_charge
+    vk3_plus ./= total_electron_charge
+    v1_plus = sum(vk1_plus)
+    v2_plus = sum(vk2_plus)
+    v3_plus = sum(vk3_plus)
+
+    return v1_plus, v2_plus, v3_plus, SVector(vk1_plus), SVector(vk2_plus),
+           SVector(vk3_plus)
+end
+
+"""
+    get_component(k, u, equations::AbstractIdealGlmMhdMultiIonEquations)
+
+Get the hydrodynamic variables of component (ion species) `k`.
+"""
+@inline function get_component(k, u, equations::AbstractIdealGlmMhdMultiIonEquations)
+    return SVector(u[3 + (k - 1) * 5 + 1],
+                   u[3 + (k - 1) * 5 + 2],
+                   u[3 + (k - 1) * 5 + 3],
+                   u[3 + (k - 1) * 5 + 4],
+                   u[3 + (k - 1) * 5 + 5])
+end
+
+"""
+    set_component!(u, k, u1, u2, u3, u4, u5,
+                   equations::AbstractIdealGlmMhdMultiIonEquations)
+
+Set the hydrodynamic variables (`u1` to `u5`) of component (ion species) `k`.
+"""
+@inline function set_component!(u, k, u1, u2, u3, u4, u5,
+                                equations::AbstractIdealGlmMhdMultiIonEquations)
+    u[3 + (k - 1) * 5 + 1] = u1
+    u[3 + (k - 1) * 5 + 2] = u2
+    u[3 + (k - 1) * 5 + 3] = u3
+    u[3 + (k - 1) * 5 + 4] = u4
+    u[3 + (k - 1) * 5 + 5] = u5
+
+    return u
+end
+
+# Extract magnetic field from solution vector
+magnetic_field(u, equations::AbstractIdealGlmMhdMultiIonEquations) = SVector(u[1], u[2],
+                                                                             u[3])
+
+# Extract GLM divergence-cleaning field from solution vector
+divergence_cleaning_field(u, equations::AbstractIdealGlmMhdMultiIonEquations) = u[end]
+
+# Get total density as the sum of the individual densities of the ion species
+@inline function density(u, equations::AbstractIdealGlmMhdMultiIonEquations)
+    rho = zero(real(equations))
+    for k in eachcomponent(equations)
+        rho += u[3 + (k - 1) * 5 + 1]
+    end
+    return rho
+end
+
+#Convert conservative variables to primitive
+function cons2prim(u, equations::AbstractIdealGlmMhdMultiIonEquations)
+    @unpack gammas = equations
+    B1, B2, B3 = magnetic_field(u, equations)
+    psi = divergence_cleaning_field(u, equations)
+
+    prim = zero(MVector{nvariables(equations), eltype(u)})
+    prim[1] = B1
+    prim[2] = B2
+    prim[3] = B3
+    for k in eachcomponent(equations)
+        rho, rho_v1, rho_v2, rho_v3, rho_e = get_component(k, u, equations)
+        srho = 1 / rho
+        v1 = srho * rho_v1
+        v2 = srho * rho_v2
+        v3 = srho * rho_v3
```

Comment:
```suggestion
        rho_inv = 1 / rho
        v1 = rho_inv * rho_v1
        v2 = rho_inv * rho_v2
        v3 = rho_inv * rho_v3
```

Reply:
Thanks. I replaced all occurrences in ee33ce176a756cfc4571009b134dfeb841c21fde.